src/Connection.ts

```diff
@@ -212,154 +167,89 @@ class Connection {
-		const bootstrapped = await Promise.all(
-			bootstrapFeatures.map(async (feature) => {
-				if (this.serverFeatureSet.validFeatures.has(feature)) {
-					return false;
-				}
-
-				this.logMessage('info', `deploying the "${feature}" feature to ${sourceDir}`);
-				const data: DbActionInputDeploy = {
-					action: 'deploy',
-					sourceDir,
-					source: await Connection.getUnibasicSource(feature),
-					programName: Connection.getServerProgramName(feature),
-				};
-
-				await this.executeDb(data);
-				return true;
-			}),
-		);
-
-		if (bootstrapped.includes(true)) {
-			// Bootstrap features needed for the deployment feature were installed, restart the deployment process
-			await this.deployFeatures(sourceDir, options);
-			return;
-		}
-
-		// deploy any other missing features
-		await Promise.all(
-			Array.from(this.serverFeatureSet.invalidFeatures).map(async (feature) => {
-				this.logMessage('info', `deploying ${feature} to ${sourceDir}`);
-				const executeDbFeatureOptions = {
-					sourceDir,
-					source: await Connection.getUnibasicSource(feature),
-					programName: Connection.getServerProgramName(feature),
-				};
-				await this.executeDbFeature('deploy', executeDbFeatureOptions);
-			}),
-		);
+	/** Deploy source code to MVIS & db server */
+	public async deploy(sourceDir: string, options?: DeployOptions): Promise<void> {
+		return this.deploymentManager.deploy(sourceDir, options);
 	}
 
-	/** Execute a database feature */
-	public async executeDbFeature<
-		TFeature extends keyof (DbSubroutineInputOptionsMap & DbSubroutineResponseTypesMap),
+	/** Execute a database subroutine */
+	public async executeDbSubroutine<
+		TSubroutineName extends keyof (DbSubroutineInputOptionsMap & DbSubroutineResponseTypesMap),
 	>(
-		feature: TFeature,
-		options: DbSubroutineInputOptionsMap[TFeature],
+		subroutineName: TSubroutineName,
+		options: DbSubroutineInputOptionsMap[TSubroutineName],
 		setupOptions: DbSubroutineSetupOptions = {},
 		teardownOptions: Record<string, never> = {},
-	): Promise<DbSubroutineResponseTypesMap[TFeature]['output']> {
-		this.logMessage('debug', `executing database feature "${feature}"`);
-
-		const featureVersion = this.serverFeatureSet.validFeatures.get(feature);
-		const setupVersion = this.serverFeatureSet.validFeatures.get('setup');
-		const teardownVersion = this.serverFeatureSet.validFeatures.get('teardown');
-
-		const data: DbActionInputSubroutine<DbSubroutineInputOptionsMap[TFeature]> = {
-			action: 'subroutine',
-			// make sure to use the compatible server version of feature
-			subroutineId: Connection.getServerProgramName(feature, featureVersion),
-			setupId: Connection.getServerProgramName('setup', setupVersion),
-			teardownId: Connection.getServerProgramName('teardown', teardownVersion),
-			options,
+	): Promise<DbSubroutineResponseTypesMap[TSubroutineName]['output']> {
+		if (this.status !== ConnectionStatus.connected) {
```

Comment:
Previously you could execute a call to the DB prior to the connection being opened because we needed to execute db subroutine functionality to confirm that the connection _could_ be opened.  This was a bit of a hole but it couldn't really be closed up too easily.  With MVIS admin doing all of the validation now, I was able to close things up so you can't execute a db subroutine until the connection has been firmly opened.